Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,504 @@
+    /// creating connection providers associated to a certain request in constant time.
+    struct Key: Hashable {
+        init(request: HTTPClient.Request) {
+            switch request.scheme {
+            case "http":
+                self.scheme = .http
+            case "https":
+                self.scheme = .https
+            default:
+                fatalError("HTTPClient.Request scheme should already be a valid one")
+            }
+            self.port = request.port
+            self.host = request.host
+        }
+
+        var scheme: Scheme
+        var host: String
+        var port: Int
+
+        enum Scheme: Hashable {
+            case http
+            case https
+        }
+    }
+
+    class Connection {
+        init(key: Key, channel: Channel, parentPool: ConnectionPool) {
+            self.key = key
+            self.channel = channel
+            self.parentPool = parentPool
+        }
+
+        func release() {
+            self.parentPool.release(self)
+        }
+
+        let parentPool: ConnectionPool
+        fileprivate let key: Key
+        let channel: Channel
+        var isLeased: Bool = false
+        var isActive: Bool {
+            return self.channel.isActive
+        }
+    }
+
+    /// Abstracts the underlying kind of connection provider
+    enum ConnectionProvider {
+        case http1(HTTP1ConnectionProvider)
+        indirect case future(EventLoopFuture<ConnectionProvider>)
+
+        func getConnection(preference: HTTPClient.EventLoopPreference) -> EventLoopFuture<Connection> {
+            switch self {
+            case .http1(let provider):
+                return provider.getConnection(preference: preference)
+            case .future(let futureProvider):
+                return futureProvider.flatMap { provider in
+                    provider.getConnection(preference: preference)
+                }
+            }
+        }
+
+        func release(connection: Connection) {
+            switch self {
+            case .http1(let provider):
+                return provider.release(connection: connection)
+            case .future(let futureProvider):
+                futureProvider.whenSuccess { provider in
+                    provider.release(connection: connection)
+                }
+            }
+        }
+
+        func syncClose() throws {
+            switch self {
+            case .http1(let provider):
+                try provider.syncClose()
+            case .future(let futureProvider):
+                try futureProvider.wait().syncClose()
+            }
+        }
+    }
+
+    class HTTP1ConnectionProvider {
+        let eventLoop: EventLoop
+        let configuration: HTTPClient.Configuration
+        let key: ConnectionPool.Key
+        private var state: State
+        var isClosed: NIOAtomic<Bool>
+        let maximumConcurrentConnections: Int = 8
+        let lock = Lock()
+        let parentPool: ConnectionPool
```

Comment:
Again, can we sort out the visibility of fields here?

Reply:
Done, thanks! Set all I could to the most restrictive visibilities possible, maybe a bit too much but I think its better to change this one at a time when a test *actually* needs to access it.